src/main/java/org/jlab/clas/std/services/convertors/EtRingToHipoReader.java

```diff
@@ -0,0 +1,293 @@
+        private EvioDataDictionary dict = null;
+        private final Queue<HipoEvent> evQueue = new LinkedList<>();
+
+        EtReader(String system, String host, int port, int runNumber,
+                 float torusField, float solenoidField) throws Exception {
+            this.runNumber = runNumber;
+            this.torusField = torusField;
+            this.solenoidField = solenoidField;
+            EtSystemOpenConfig config = new EtSystemOpenConfig(system, host, port);
+            config.setConnectRemotely(true);
+            sys = new EtSystem(config, EtConstants.debugInfo);
+            sys.open();
+
+            EtStationConfig statConfig = new EtStationConfig();
+            statConfig.setFlowMode(EtConstants.stationSerial);
+            statConfig.setBlockMode(EtConstants.stationNonBlocking);
+            statConfig.setCue(QSIZE);
+            dict = new EvioDataDictionary();
+            dict.initWithEnv("CLAS12DIR", "/etc/bankdefs/clas12");
+            EtStation stat = sys.createStation(statConfig, DEFAULT_STAT_NAME);
+            att = sys.attach(stat);
+
+            System.out.println("INFO: Start processing with runNumber = " + runNumber
+                + ", torus = " + this.torusField
+                + ", solenoid = " + this.solenoidField);
+        }
+
+        HipoEvent getEvent() throws Exception {
+            if (evQueue.isEmpty()) {
+                getEtEvents();
+            }
+            return evQueue.poll();
+        }
+
+        private void getEtEvents() throws Exception {
+
+            EtEvent[] mevs;
+            try {
+
+
+                // get events from ET system
+                try {
+                    mevs = sys.getEvents(att, Mode.SLEEP, null, TIMEOUT, MAX_NEVENTS);
+//                    mevs = sys.getEvents(att, Mode.TIMED, null, TIMEOUT, MAX_NEVENTS);
+                } catch (EtTimeoutException e) {
+                    System.out.println("Timed out, try again");
+                    return;
+                }
+
+                // ----------- ET
+                for (EtEvent mev : mevs) {
+                    // Get event's data buffer
+                    // buf.limit() = length of the actual data (not buffer capacity)
+                    ByteBuffer buf = mev.getDataBuffer();
+                    buf.limit(mev.getLength()).position(0);
+
+//                    Utilities.printBuffer(buf,0, 30, "et-data");
+//                    System.out.println("    data len = " + mev.getLength());
+
+                    try {
+                        // If using byte array you need to watch out for endianness
+                        ByteBuffer evioBuffer = ByteBuffer.allocate(mev.getLength());
+                        try {
+                            byte[] data = mev.getData();
+                            evioBuffer.put(data, 0, mev.getLength());
+                        } catch (UnsupportedOperationException e) {
+                            evioBuffer.put(buf);
+                        }
```

Comment:
Is using an exception as control flow the only option? There could be a way to check the existence of the array.

Reply:
I stand by my previous answer.